### PR TITLE
Chained PMDAO

### PR DESCRIPTION
**Pull request by michaelconnor468**
created pipelined PMDAO classes

### CODE REVIEW

File: src/foam/dao/EasyDAO.js [hunk 1]
```diff
@@ -100,38 +100,116 @@ foam.dao.DAO delegate = getInnerDAO() == null ?
 
 if ( delegate instanceof foam.dao.MDAO ) setMdao((foam.dao.MDAO)delegate);
 
+PipelinedPMDAO pipelineDAO = null;
+
+if(getPipelinedPMDAO()){
```
Comment (kgrgreer) [inline src/foam/dao/EasyDAO.js]: FOAM coding standard is:
if ( getPipelinedPMDAO() ) {
See: https://github.com/foam-framework/foam/wiki/CodingStyle

File: src/foam/dao/EasyDAO.js [hunk 2]
```diff
@@ -100,38 +100,116 @@ foam.dao.DAO delegate = getInnerDAO() == null ?
...
+  pipelineDAO = new foam.dao.PipelinedPMDAO(getX(), delegate);
+  delegate = pipelineDAO;
+}
+
 if ( getJournalType().equals(JournalType.SINGLE_JOURNAL) ) {
   delegate = new foam.dao.java.JDAO(getX(), delegate, getJournalName());
 }
 
+if(getPipelinedPMDAO()){
```
Comment (kgrgreer) [inline src/foam/dao/EasyDAO.js]: Why are you repeating this check 8 times?

File: src/foam/dao/PipelinedPMDAO.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/dao/PipelinedPMDAO.java]: 2019

File: src/foam/dao/EasyDAO.js [hunk 4]
```diff
@@ -144,6 +163,12 @@ return delegate;
       type: 'foam.dao.DAO',
       name: 'innerDAO'
     },
+    {
+      /** Have EasyDAO create pipeline of PMs which measure the performance of each node in the decorator chain */
```
Comment (kgrgreer) [inline src/foam/dao/EasyDAO.js:167]: documentation: 'Have EasyDAO...'

File: src/foam/dao/EasyDAO.js [hunk 5]
```diff
@@ -144,6 +163,12 @@ return delegate;
...
+      class: 'Boolean',
+      name: 'pipelinedPMDAO',
+      value: false
```
Comment (kgrgreer) [inline src/foam/dao/EasyDAO.js:170]: false is the default default value

File: src/foam/dao/PipelinedPMDAO.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+public class PipelinedPMDAO
+  extends ProxyDAO
+{
+
+  protected PM chainPM;
+  protected PipelinedPMDAO chainStart_;
+  
+    public PipelinedPMDAO(X x, DAO delegate) {
+        super(x, delegate);
```
Comment (kgrgreer) [inline src/foam/dao/PipelinedPMDAO.java:23]: indent two spaces

File: src/foam/dao/PipelinedPMDAO.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import foam.nanos.pm.PM;
+
+public class PipelinedPMDAO
+  extends ProxyDAO
+{
+
+  protected PM chainPM;
+  protected PipelinedPMDAO chainStart_;
+  
+    public PipelinedPMDAO(X x, DAO delegate) {
```
Comment (kgrgreer) [inline src/foam/dao/PipelinedPMDAO.java:22]: indent two spaces

**Comment by jlhughes** [comment]
Replaced by modeled  class.

**Pull request closed without merging**